chia/server/upnp.py

```diff
@@ -9,14 +11,52 @@
 log = logging.getLogger(__name__)
 
 
-def upnp_remap_port(port) -> None:
-    log.info(f"Attempting to enable UPnP (open up port {port})")
-    try:
-        upnp = miniupnpc.UPnP()
-        upnp.discoverdelay = 30
-        upnp.discover()
-        upnp.selectigd()
-        upnp.addportmapping(port, "TCP", upnp.lanaddr, port, "chia", "")
-        log.info(f"Port {port} opened with UPnP. lanaddr {upnp.lanaddr} external: {upnp.externalipaddress()}")
-    except Exception:
-        log.info("UPnP failed. This is not required to run chia, but it allows incoming connections from other peers.")
+class UPnP:
+    def __init__(self):
+        self.queue = Queue()
+
+        def run():
+            try:
+                self.upnp = miniupnpc.UPnP()
+                self.upnp.discoverdelay = 30
+                self.upnp.discover()
+                self.upnp.selectigd()
+                keep_going = True
+                while keep_going:
+                    msg = self.queue.get()
+                    if msg[0] == 'remap':
+                        port = msg[1]
+                        log.info(f"Attempting to enable UPnP (open up port {port})")
+                        self.upnp.deleteportmapping(port, "TCP")
+                        self.upnp.addportmapping(port, "TCP", self.upnp.lanaddr, port, "chia", "")
+                        log.info(f"Port {port} opened with UPnP. lanaddr {self.upnp.lanaddr} "
+                                 f"external: {self.upnp.externalipaddress()}")
+                    elif msg[0] == 'release':
+                        port = msg[1]
+                        self.upnp.deleteportmapping(port, "TCP")
+                        log.info(f"Port {port} closed with UPnP")
+                    elif msg[0] == 'shutdown':
+                        keep_going = False
+            except Exception as e:
+                log.info(
+                    "UPnP failed. This is not required to run chia, it allows incoming connections from other peers.")
+                log.info(e)
+
+        self.thread = threading.Thread(target=run)
+        self.thread.start()
+
+    def remap(self, port):
+        self.queue.put(('remap', port))
+
+    def release(self, port):
+        self.queue.put(('release', port))
+
+    def shutdown(self):
+        self.queue.put(('shutdown',))
+        self.thread.join()
+
+    def __del__(self):
```

Comment:
when is this called exactly? It would be nice if we could `await` the thread shutting down, but that's a luxury feature. This patch seems like a great improvement as is

Reply:
I added it so that we could shut it down cleanly if needed by either destroying the object, or explicitly calling shutdown. In chia's lifecycle the only time we are done with upnp is right before the process is torn down anyway, so running the shutdown code doesn't seem to serve a purpose. If I'm wrong about this though, it's ready to use.

Reply:
this is the only place we call `shutdown()` though, right? if this isn't called, the thread will outlive the process and terminate non-gracefully, and possibly leave the UPnP port open